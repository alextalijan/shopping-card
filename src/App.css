.header {
  background-color: #1e293b;
  color: white;
  height: 100px;
  display: flex;
  justify-content: end;
  align-items: center;
  gap: 2rem;
}

.nav-link {
  text-decoration: none;
  font-size: 2rem;
  transition: transform 0.3s;
}

.nav-link:hover {
  transform: scale(1.2);
}

.nav-link:visited {
  color: white;
}

.nav-link:last-child {
  margin-right: 2rem;
}
