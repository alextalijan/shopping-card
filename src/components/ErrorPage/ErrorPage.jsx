import { Link } from 'react-router-dom';
import styles from './ErrorPage.module.css';

function ErrorPage() {
  return (
    <>
      <h1 className={styles.h1}>Error: Page Not Found</h1>
      <p className={styles['error-message']} data-testid="error-text">
        Make sure you've requested an existing page.
      </p>
      <p className={styles['error-message']} data-testid="error-text">
        Click to go back to the{' '}
        <Link className={styles['nav-link']} to="/">
          home page
        </Link>
        .
      </p>
      <img className={styles['error-img']} src="/public/error.png" alt="" />
    </>
  );
}

export default ErrorPage;
