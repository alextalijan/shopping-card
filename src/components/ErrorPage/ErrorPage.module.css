.h1 {
  font-size: 4rem;
  text-align: center;
  margin: 8rem auto 3rem;
}

.error-message {
  font-size: 2rem;
  text-align: center;
  margin: 2rem auto;
  padding: 0 2rem;
}

.nav-link {
  text-decoration: none;
  color: blue;
}

.error-img {
  width: 200px;
  display: block;
  margin: 0 auto;
}

@media (max-width: 720px) {
  .h1 {
    font-size: 3.5rem;
  }

  .error-message {
    font-size: 1.7rem;
  }
}
