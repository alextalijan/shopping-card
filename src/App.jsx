import { useEffect, useState } from 'react';
import './App.css';
import { Outlet, Link } from 'react-router-dom';

function App() {
  const [cart, setCart] = useState([]);
  const [shopItems, setShopItems] = useState([]);
  const [loadingShop, setLoadingShop] = useState(true);
  const [shopError, setShopError] = useState(null);

  function addToCart(id, title, imageSrc, price, amount) {
    if (amount === 0) return;

    const addedItem = {
      id,
      title,
      imageSrc,
      price,
      amount,
    };

    for (const item of cart) {
      if (addedItem.id === item.id) {
        const newCart = cart.map((item) => {
          if (item.id === addedItem.id) {
            return {
              id: addedItem.id,
              title: addedItem.title,
              price: addedItem.price,
              amount: item.amount + addedItem.amount,
            };
          }

          return item;
        });

        setCart(newCart);
        return;
      }
    }

    setCart([...cart, addedItem]);
  }

  function removeFromCart(id) {
    setCart(cart.filter((item) => item.id !== id));
  }

  useEffect(() => {
    fetch('https://fakestoreapi.com/products/')
      .then((response) => {
        if (!response.ok) {
          throw new Error(
            'Failed to load items from the shop. Please try again by refreshing the page.',
          );
        }

        return response.json();
      })
      .then((data) => {
        const items = [];

        for (let i = 0; i < 10; i += 1) {
          items.push({
            id: data[i].id,
            title: data[i].title,
            price: data[i].price,
            imageSrc: data[i].image,
          });
        }

        setShopItems(items);
      })
      .catch((error) => {
        setShopError(error.message);
      })
      .finally(() => {
        setLoadingShop(false);
      });
  }, []);

  return (
    <>
      <header className="header">
        <Link className="nav-link" to="/">
          Home
        </Link>
        <Link className="nav-link" to="/shop">
          Shop
        </Link>
        <Link className="nav-link" to="/cart">
          Cart <span>({cart.length})</span>
        </Link>
      </header>
      <main>
        <Outlet
          context={{
            shopItems,
            loadingShop,
            shopError,
            addToCart,
            cart,
            setCart,
            removeFromCart,
          }}
        />
      </main>
    </>
  );
}

export default App;
