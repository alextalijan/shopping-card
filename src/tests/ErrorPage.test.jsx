import { render, screen } from '@testing-library/react';
import { it, expect } from 'vitest';
import ErrorPage from '../components/ErrorPage/ErrorPage';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import App from '../App';

it('renders', () => {
  const { container } = render(
    <MemoryRouter>
      <ErrorPage />
    </MemoryRouter>,
  );

  expect(container).toMatchSnapshot();
});

it('shows the error page when a non-existing route is provided', () => {
  render(
    <MemoryRouter initialEntries={['/non-existing-page']}>
      <Routes>
        <Route path="/" element={<App />}>
          <Route path="*" element={<ErrorPage />} />
        </Route>
      </Routes>
    </MemoryRouter>,
  );

  expect(
    screen.getByRole('heading', { name: /error: page not found/i }),
  ).toBeInTheDocument();
  expect(screen.getAllByTestId('error-text').length).toBe(2);
  expect(screen.getByAltText('')).toBeInTheDocument();
});
