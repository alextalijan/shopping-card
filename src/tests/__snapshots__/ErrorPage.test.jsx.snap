// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders 1`] = `
<div>
  <h1
    class="_h1_955fbf"
  >
    Error: Page Not Found
  </h1>
  <p
    class="_error-message_955fbf"
    data-testid="error-text"
  >
    Make sure you've requested an existing page.
  </p>
  <p
    class="_error-message_955fbf"
    data-testid="error-text"
  >
    Click to go back to the
     
    <a
      class="_nav-link_955fbf"
      data-discover="true"
      href="/"
    >
      home page
    </a>
    .
  </p>
  <img
    alt=""
    class="_error-img_955fbf"
    src="/public/error.png"
  />
</div>
`;
